### PR TITLE
add BinTransform, TimeUnitTransform, and AggregateTransform

### LINKED ISSUE #2344: Add bin, aggregate, timeUnit to the transform array
The `transform` array should support the following transforms:

- `BinTransform`

```json
{
  "bin": true, "field": "acceleration", "as": "binned_acceleration"
}
```

- `TimeUnitTransform`

```json
{
  "timeUnit": "year", "field": "date", "as": "year_date"
}
```

- `AggregateTransform`

```json
{
  "summarize": [
    {"aggregate": "sum", "field": "acceleration", "as": "sum_acceleration"},
    {"aggregate": "mean", "field": "horsepower", "as": "mean_horsepower"}
  ],
  "groupby": ["origin", "cylinders"]
}
```

## TODO 

- [ ] Add these interfaces to `transform.ts`
- [ ] Modify `parseTransformArray` in `src/compile/data/transforms.ts`
All of these transforms should be able to reuse existing `Bin/TimeUnit/AggregateNode` in the transform pipeline. 

- [ ] Add a test example where we calculate / filter based on aggregate value
  - [ ] Create a "full" example of boxplot with 1.5 IQR whiskers

### PR DESCRIPTION
**Pull request by YuhanLu**
Fix #2344

### CODE REVIEW

File: src/compile/data/transforms.ts [hunk 1]
```diff
@@ -65,6 +66,9 @@ export function parseTransformArray(model: Model) {
       node = new CalculateNode(t);
     } else if (isFilter(t)) {
       node = new FilterNode(model, t.filter);
+    } else if (isBin(t)) {
+      const modelWithField = model as ModelWithField;
+      node = BinNode.makeTransform(modelWithField, t);
```
Comment (kanitw) [inline src/compile/data/transforms.ts]: 1) As mentioned a few times earlier, in general, you should never cast anything in the main code without a good reason.

2) In this case, I don't see why the model here should be a `ModelWithField` -- it should just be `Model`.

File: src/transform.ts [hunk 2]
```diff
@@ -23,8 +27,74 @@ export interface CalculateTransform {
...
+   */
+  field: string;
+
+  /**
+   * The field for storing the computed formula value.
+   */
+  as: string;
+}
+
+export interface AggregateTransform {
```
Comment (kanitw) [inline src/transform.ts]: `SummarizeTransform`

File: src/transform.ts [hunk 3]
```diff
@@ -23,8 +27,74 @@ export interface CalculateTransform {
...
+   */
+  summarize: Aggregrate[];
+
+  /**
+   * Array of fields we will be useing for group by
+   */
+  groupby: VgFieldRef[];
+}
+
+export interface Aggregrate {
```
Comment (kanitw) [inline src/transform.ts]: `Summarize`

File: test/compile/data/bin.test.ts [hunk 4]
```diff
@@ -4,14 +4,15 @@ import {assert} from 'chai';
...
+import {BinTransform} from '../../../src/transform';
 import {VgTransform} from '../../../src/vega.schema';
 import {parseUnitModel} from '../../util';
 
 function assemble(model: ModelWithField) {
   return BinNode.make(model).assemble();
 }
 
-describe('compile/data/bin', function() {
+describe.only('compile/data/bin', function() {
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: Remove "only"

File: test/compile/data/bin.test.ts [hunk 5]
```diff
@@ -102,4 +103,64 @@ describe('compile/data/bin', function() {
       expr: 'format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_start"], \'s\') + \' - \' + format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_end"], \'s\')'
     });
   });
+
+  it ('should add bin transform and correctly apply bin for binned field without custom extent', () => {
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: Are you just adding this to improve test coverage? 
(It seems irrelevant to what you are adding in this PR)

File: test/compile/data/bin.test.ts [hunk 6]
```diff
@@ -102,4 +103,64 @@ describe('compile/data/bin', function() {
...
+    });
+  });
+
+  it('should apply the bin transform if specified BinTransform', () => {
+    const t: BinTransform = {bin: {"maxbins" : 10}, field: "Rotten_Tomatoes_Rating", as: "something"};
+    const model = parseUnitModel({
+      data: {url: "data/movies.json"},
+      mark: "circle",
+      transform: [t],
+      encoding: {
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: Not sure why you need to add x and color here -- in any case this is a funny spec because you don't really use the output field from `BinTransform`.

File: test/compile/data/bin.test.ts [hunk 7]
```diff
@@ -102,4 +103,64 @@ describe('compile/data/bin', function() {
...
+      field: 'Acceleration',
+      as: ['bin_maxbins_10_Acceleration_start', 'bin_maxbins_10_Acceleration_end'],
+      maxbins: 10,
+      signal: 'bin_maxbins_10_Acceleration_bins',
+      extent: {signal: 'bin_maxbins_10_Acceleration_extent'}
+    });
+  });
+
+  it('should apply the bin transform if specified BinTransform', () => {
+    const t: BinTransform = {bin: {"maxbins" : 10}, field: "Rotten_Tomatoes_Rating", as: "something"};
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: Can you name it a bit better than something

File: src/compile/data/bin.ts [hunk 8]
```diff
@@ -89,6 +90,38 @@ export class BinNode extends DataFlowNode {
     return new BinNode(bins);
   }
 
+  public static makeTransform(model: ModelWithField, t: BinTransform) {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: As mentioned in person, you can name this `makeBinFromTransform` and rename the other make to `makeBinFromEncoding`.

File: src/compile/data/bin.ts [hunk 9]
```diff
@@ -89,6 +90,38 @@ export class BinNode extends DataFlowNode {
...
+    const bins: Dict<BinComponent> = {};
+    model.forEachFieldDef((fieldDef, channel) => {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: Why is this calling `forEachFieldDef`, which basically enumerates `encoding`?  

This bin in the transform basically should be orthogonal to the encoding.

File: src/compile/data/bin.ts [hunk 10]
```diff
@@ -89,6 +90,27 @@ export class BinNode extends DataFlowNode {
     return new BinNode(bins);
   }
 
+  public static makeBinFromTransform(model: Model, t: BinTransform) {
+    const bins: Dict<BinComponent> = {};
+
+    if (t.bin) {
+      const bin: Bin = isBoolean(t.bin) ? {} : t.bin;
+      const key = `${binToString(t.bin)}_${t.field}`;
```
Comment (kanitw) [inline src/compile/data/bin.ts]: We need to augment `bin` with `maxbins` if both `maxbins` and step` are` not provided like in `normalize` in fielddef.ts

Comment (kanitw) [inline src/compile/data/bin.ts]: (You might wanna extract `normalizeBin` method from `normalize` in `fielddef.ts` and put that in `bin.ts` so that we can reuse them in both places)

File: src/compile/data/bin.ts [hunk 11]
```diff
@@ -89,6 +90,27 @@ export class BinNode extends DataFlowNode {
...
+      bins[key] = {
+        bin: bin,
+        field: t.field,
+        as: [binToString(t.bin) + '_' + t.field + '_start', binToString(t.bin) + '_' + t.field + '_end'],
```
Comment (kanitw) [inline src/compile/data/bin.ts]: Why is this different from the logic from binFromEncoding?

File: src/compile/data/bin.ts [hunk 12]
```diff
@@ -89,6 +90,27 @@ export class BinNode extends DataFlowNode {
     return new BinNode(bins);
   }
 
+  public static makeBinFromTransform(model: Model, t: BinTransform) {
+    const bins: Dict<BinComponent> = {};
+
+    if (t.bin) {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: Can `t.bin` really be falsy here?

File: src/compile/data/bin.ts [hunk 13]
```diff
@@ -89,6 +90,27 @@ export class BinNode extends DataFlowNode {
...
+      const key = `${binToString(t.bin)}_${t.field}`;
+      bins[key] = {
+        bin: bin,
+        field: t.field,
+        as: [binToString(t.bin) + '_' + t.field + '_start', binToString(t.bin) + '_' + t.field + '_end'],
+        signal: model.getName(`${key}_bins`),
+        extentSignal: model.getName(key + '_extent')
+      };
+    }
+    if (Object.keys(bins).length === 0) {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: Can the length really be 0 here?

File: src/compile/data/bin.ts [hunk 14]
```diff
@@ -89,6 +90,27 @@ export class BinNode extends DataFlowNode {
     return new BinNode(bins);
   }
 
+  public static makeBinFromTransform(model: Model, t: BinTransform) {
+    const bins: Dict<BinComponent> = {};
+
+    if (t.bin) {
+      const bin: Bin = isBoolean(t.bin) ? {} : t.bin;
+      const key = `${binToString(t.bin)}_${t.field}`;
+      bins[key] = {
```
Comment (kanitw) [inline src/compile/data/bin.ts]: You might want to extract a method so that we use common method to create the object here.

File: src/compile/data/timeunit.ts [hunk 15]
```diff
@@ -44,6 +45,24 @@ export class TimeUnitNode extends DataFlowNode {
     return new TimeUnitNode(formula);
   }
 
+  public static makeFromTransfrom(model: Model, t: TimeUnitTransform) {
+    const timeUnitComponent: Dict<TimeUnitComponent> = {};
+      if (t.timeUnit) {
```
Comment (kanitw) [inline src/compile/data/timeunit.ts]: 1) Why tabbed?
2) Can TimeUnit really be falsy here?

File: src/compile/data/timeunit.ts [hunk 16]
```diff
@@ -44,6 +45,24 @@ export class TimeUnitNode extends DataFlowNode {
...
+        timeUnitComponent[t.field] = {
+          as: t.as,
+          timeUnit: t.timeUnit,
+          field: t.field
+        };
+      }
+
+
+    if (Object.keys(timeUnitComponent).length === 0) {
+      return null;
```
Comment (kanitw) [inline src/compile/data/timeunit.ts]: Can the object really be empty here?

File: src/compile/data/transforms.ts [hunk 17]
```diff
@@ -65,6 +68,12 @@ export function parseTransformArray(model: Model) {
       node = new CalculateNode(t);
     } else if (isFilter(t)) {
       node = new FilterNode(model, t.filter);
+    } else if (isBin(t)) {
```
Comment (kanitw) [inline src/compile/data/transforms.ts]: Create a test that include all types of transform to test this method?

File: test/compile/data/aggregate.test.ts [hunk 18]
```diff
@@ -114,5 +115,29 @@ describe('compile/data/summary', function () {
         fields: ['*']
       });
     });
+
+    it('should produce the correct summary component from transform array', function() {
+      const t: SummarizeTransform = {summarize: [
```
Comment (kanitw) [inline test/compile/data/aggregate.test.ts]: tab better

File: test/compile/data/bin.test.ts [hunk 19]
```diff
@@ -102,4 +107,33 @@ describe('compile/data/bin', function() {
       expr: 'format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_start"], \'s\') + \' - \' + format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_end"], \'s\')'
     });
   });
+
+  it('should add bin transform from transform arrat and correctly apply bin with custom extent', function() {
+    const t: BinTransform = {bin: {extent: [0, 100]},
+                  field: 'Acceleration',
+                  as: 'Bin_Transform'};
+     const model = parseUnitModel({
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: not align

File: test/compile/data/bin.test.ts [hunk 20]
```diff
@@ -102,4 +107,33 @@ describe('compile/data/bin', function() {
       expr: 'format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_start"], \'s\') + \' - \' + format(datum["bin_maxbins_10_Rotten_Tomatoes_Rating_end"], \'s\')'
     });
   });
+
+  it('should add bin transform from transform arrat and correctly apply bin with custom extent', function() {
+    const t: BinTransform = {bin: {extent: [0, 100]},
```
Comment (kanitw) [inline test/compile/data/bin.test.ts]: bad tab

File: src/compile/data/aggregate.ts [hunk 21]
```diff
@@ -99,7 +101,36 @@ export class AggregateNode extends DataFlowNode {
...
+    for(const s of t.summarize) {
+      if (s.aggregate) {
+        if (s.aggregate === 'count') {
+          meas['*'] = meas['*'] || {};
+          meas['*']['count'] = true;
+          as['*'] = s.as;
+        } else {
+          meas[s.field] = meas[s.field] || {};
+          meas[s.field][s.aggregate] = true;
+          as[s.field] = s.as;
```
Comment (kanitw) [inline src/compile/data/aggregate.ts]: Or even just `Dict<Dict<string>>` and always output name.

File: src/compile/data/bin.ts (hunk 11 above)
Comment (YuhanLu) [inline src/compile/data/bin.ts]: I manually extracted the correct logic from what we have in the other method without encoding channel

File: src/compile/data/bin.ts [hunk 22]
```diff
@@ -93,20 +97,7 @@ export class BinNode extends DataFlowNode {
...
-        field: t.field,
-        as: [binToString(t.bin) + '_' + t.field + '_start', binToString(t.bin) + '_' + t.field + '_end'],
-        signal: model.getName(`${key}_bins`),
-        extentSignal: model.getName(key + '_extent')
-      };
-    }
-    if (Object.keys(bins).length === 0) {
-      return null;
-    }
+    createBin(bins, t, model, undefined);
```
Comment (kanitw) [inline src/compile/data/bin.ts]:
```
const key = binKey(...);

return new BinNode({
  [key]: createBinComponent(...)
});
```

File: src/compile/data/timeunit.ts [hunk 23]
```diff
@@ -47,18 +47,11 @@ export class TimeUnitNode extends DataFlowNode {
...
-          timeUnit: t.timeUnit,
-          field: t.field
-        };
-      }
-
-
-    if (Object.keys(timeUnitComponent).length === 0) {
-      return null;
-    }
+    timeUnitComponent[t.field] = {
```
Comment (kanitw) [inline src/compile/data/timeunit.ts]: Just make this one line
```
return new TimeUnitNode({
  [t.field]: ...
}
```

File: src/fielddef.ts [hunk 24]
```diff
@@ -285,22 +286,9 @@ export function normalize(channelDef: ChannelDef<string>, channel: Channel) {
...
-      } else if (!bin.maxbins && !bin.step) {
-        fieldDef = {
-          ...fieldDef,
-          bin: {
-            ...bin,
-            maxbins: autoMaxBins(channel)
-          }
-        };
-      }
+    const bin = normalizeBin(fieldDef.bin, channel);
```
Comment (kanitw) [inline src/fielddef.ts]: 1) Why should you call this no matter fieldDef.bin is truthy or not?  The predicate for if should be `fieldDef.bin` and `normalizeBin` should be called inside.  
2) Setting `fieldDef.bin` cause side effect.  Why don't you follow the pattern we had earlier?

File: src/fielddef.ts [hunk 25]
```diff
@@ -332,6 +320,18 @@ export function normalize(channelDef: ChannelDef<string>, channel: Channel) {
   return channelDef;
 }
 
+export function normalizeBin(bin: Bin|boolean, channel: Channel) {
+  if (bin) {
```
Comment (kanitw) [inline src/fielddef.ts]: why `if (bin)` here?

File: test/compile/data/aggregate.test.ts (hunk 18 above)
Comment (kanitw) [inline test/compile/data/aggregate.test.ts]: @YuhanLu  This bad tab is still here...

File: test/compile/data/transform.test.ts [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+      const result = parseTransformArray(model);
+      assert.isTrue(result.first instanceof CalculateNode);
+      assert.isTrue(result.last instanceof FilterNode);
+    });
+
+    it('should return a BinNode node and a TimeUnitNode', () => {
+      const model = parseUnitModel({
+        data: {values: []},
+        mark: 'point',
+        transform: [{bin: true, field: 'field', as: 'as'}, {timeUnit: 'month', field: 'field', as: 'as'}],
```
Comment (kanitw) [inline test/compile/data/transform.test.ts]: This is a bad example -- why name two different things with the same name (`'as'`) ?

File: test/compile/data/transform.test.ts [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+      const result = parseTransformArray(model);
+      assert.isTrue(result.first instanceof BinNode);
+      assert.isTrue(result.last instanceof TimeUnitNode);
+    });
+
+    it('should return a BinNode and a AggregateNode', () => {
+      const model = parseUnitModel({
+        data: {values: []},
+        mark: 'point',
+        transform: [{bin: true, field: 'field', as: 'as'}, {summarize: [{aggregate: 'count', field: 'f', as: 'a'}, {aggregate: 'sum', field: 'f', as: 'a'}], groupby: ['field']}],
```
Comment (kanitw) [inline test/compile/data/transform.test.ts]: This is a bad example -- why name two different things with the same name ('a') ?

File: test/compile/data/transform.test.ts [hunk 28]
```diff
@@ -0,0 +1,55 @@
+import {assert} from 'chai';
+import {AggregateNode} from '../../../src/compile/data/aggregate';
+import {BinNode} from '../../../src/compile/data/bin';
+import {TimeUnitNode} from '../../../src/compile/data/timeunit';
+import {CalculateNode, FilterNode, parseTransformArray} from '../../../src/compile/data/transforms';
+import {TimeUnit} from '../../../src/timeunit';
+import {parseUnitModel} from '../../util';
+describe('compile/data/transform', () => {
```
Comment (kanitw) [inline test/compile/data/transform.test.ts]: add a line break between import and describe.

File: src/compile/data/aggregate.ts [hunk 29]
```diff
@@ -117,7 +117,8 @@ export class AggregateNode extends DataFlowNode {
         } else {
           meas[s.field] = meas[s.field] || {};
           meas[s.field][s.aggregate] = true;
-          as[s.field] = s.as;
+          as[s.field] = as[s.field] || {};
```
Comment (kanitw) [inline src/compile/data/aggregate.ts]: Remove `as` -- make `meas` a `Dict<Dict<string>>` and always include the correct as. 
(For encoding's aggregate, call `field(...)` with the right params.

Comment (kanitw) [inline src/compile/data/aggregate.ts]: For summarize transform, automatically add `as` if `as` is not defined (using `||`)

File: src/fielddef.ts (hunk 24 above)
Comment (YuhanLu) [inline src/fielddef.ts]: Why setting fieldDef.bin will cause side effect? I see what we had before is actually overwriting fieldDef.bin, isn't it?

File: src/compile/data/transforms.ts (hunk 17 above)
Comment (YuhanLu) [inline src/compile/data/transforms.ts]: Done

File: src/compile/data/aggregate.ts [hunk 30]
```diff
@@ -26,18 +26,18 @@ function addDimension(dims: {[field: string]: boolean}, fieldDef: FieldDef<strin
...
       // when we merge a measure, we either have to add an aggregation operator or even a new field
-      const ops = childMeasures[field];
+      const ops = childMeasures[f];
       for (const op in ops) {
         if (ops.hasOwnProperty(op)) {
-          if (field in parentMeasures) {
+          if (f in parentMeasures) {
             // add operator to existing measure field
-            parentMeasures[field][op] = true;
+            parentMeasures[f][op] = field({field: f});
```
Comment (kanitw) [inline src/compile/data/aggregate.ts]: Don't call `field`

File: src/compile/data/aggregate.ts [hunk 31]
```diff
@@ -26,18 +26,18 @@ function addDimension(dims: {[field: string]: boolean}, fieldDef: FieldDef<strin
...
           } else {
-            parentMeasures[field] = {op: true};
+            parentMeasures[f] = {op: field({field: f})};
```
Comment (kanitw) [inline src/compile/data/aggregate.ts]: Don't call `field`

File: src/fielddef.ts [hunk 32]
```diff
@@ -175,7 +176,7 @@ export interface FieldRefOption {
   aggregate?: AggregateOp;
 }
 
-export function field(fieldDef: FieldDef<string>, opt: FieldRefOption = {}): string {
+export function field(fieldDef: FieldDef<string>|BinTransform, opt: FieldRefOption = {}): string {
```
Comment (kanitw) [inline src/fielddef.ts]: I already remove `|BinTransform` for you -- why you bring it back?

File: src/compile/data/aggregate.ts [hunk 33]
```diff
@@ -139,18 +165,25 @@ export class AggregateNode extends DataFlowNode {
...
 
-    return {
+    let result: VgAggregateTransform = {
       type: 'aggregate',
       groupby: keys(this.dimensions),
       ops,
       fields
     };
+
+    if (as.length !== 0) {
```
Comment (kanitw) [inline src/compile/data/aggregate.ts]: I don't think as.length can really be 0

File: src/compile/data/aggregate.ts (hunk 30 above)
Comment (kanitw) [inline src/compile/data/aggregate.ts]: I guess `ops[op]` is the right value here.

One potential issue here is that if `parentMeasures[f][op] ` already exist, and `ops[op]` is different, the merge will be incorrect.  The right solution is to make measure `Dict<Dict<string>>` -- but that's hairy -- feel free to file a bug  with very descriptive description.

File: src/compile/data/aggregate.ts (hunk 31 above)
Comment (kanitw) [inline src/compile/data/aggregate.ts]: See https://github.com/vega/vega-lite/pull/2360#discussion_r116889012

**Comment by kanitw** [comment]
One last thing -- create example.

**Comment by kanitw** [comment]
One last thing -- add example :)

File: examples/vg-specs/area.vg.json [hunk 34]
```diff
@@ -33,6 +33,9 @@
                     ],
                     "fields": [
                         "count"
+                    ],
+                    "as": [
```
Comment (domoritz) [inline examples/vg-specs/area.vg.json]: Why do we add this all of a sudden? Is it necessary? I think this should only be generated if we don't use the default.

Comment (kanitw) [inline examples/vg-specs/area.vg.json]: 1. This makes the code way more maintainable
2. This makes the output code explicit about field names, which is not necessary a bad thing.

Comment (domoritz) [inline examples/vg-specs/area.vg.json]: I was just wondering why this is on this PR. Seems a bit unrelated but whatever.

Comment (kanitw) [inline examples/vg-specs/area.vg.json]: (It's way more maintainable because AggregateNode from encoding become way easier to merge with AggregateNode from transform)

**Comment by kanitw** [comment]
@mattwchun FYI this is merged.
